# Cariaco plasmids

This repository contains scripts, data, and computing environments used for the analysis of plasmids detected in MAGs recovered from the Cariaco Basin, Venezuela for the publication "[INSERT HERE]". The authors of this paper are P. Mara, D. Geller-McGrath, G. Taylor, E. Suter, V. Edgcomb, and M. Pachiadaki. The scripts, files, and data here were written and uploaded by D. Geller-McGrath.

To view the code used for this analysis, click here: https://d-mcgrath.github.io/cariaco_plasmids/. This is a GitHub pages document that contains R scripts for figure creation and data analysis, as well as bash scripts for data analysis that was done using a high performance computing cluster (HPC). The R script chunks contain in order the code for creation of figures [INSERT HERE] in the main text, and supplementary figures [INSERT HERE]. Bash scripts for analysis done on an HPC are below the R scripts. Several Conda environments required to reproduce these HPC analyses are in the [INSERT HERE] folder. Data required to re-run the R scripts are available in the [INSERT HERE] folder.

All MAGs recovered in this study and biosynthetic gene clusters recovered from those MAGs using antiSMASH 6.0 are available in the repository: https://osf.io/usm8r/ in the folders "Prokaryotic_CarAnox_MAGs" and "Biosynthetic_gene_clusters". The raw metagenomic and metatranscriptomic sequencing reads used in this study can be accessed on [NCBI](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA326482).
